File: LuaCode/mod/NetObserver.lua
--=============================================================================
-- Copyright (C) , 金七情(Loong) jinqiqing@qq.com
-- Created by Loong on 1/15/2019, 11:27:53 AM
--=============================================================================

NetObserver = {Name="NetObserver"}

local My = NetObserver
-- ViaCarrier:移动网络
NetType = {None=0, ViaCarrier=1, Wifi=2}

function My:Init()
    self:SetType()
    self.eChange = Event()
    EventMgr.Add("NetChange",EventHandler(self.Change, self))
end

function My:SetType()
    local type = Device.NetType
    if type == "wifi" then
        self.type = NetType.Wifi
    elseif type == "4g" then
        self.type = NetType.ViaCarrier
    elseif type == "3g" then
        self.type = NetType.ViaCarrier
    elseif type == "2g" then
        self.type = NetType.ViaCarrier
    else
        self.type = NetType.None
    end
end

function My:Change(last,cur)
    self.eChange(last, cur)
end

function My:Clear()
    
end

return My

File: Pro/Assets/Script/Client/Network/Basic/NetObserver.cs
/*=============================================================================
 * Copyright (C) 2014, 金七情(Loong) jinqiqing@qq.com
 * Created by Loong in 2014.4.26 22:38:16
 * 网络状态发生改变时,会触发change事件
 ============================================================================*/

using System;
using UnityEngine;
using System.Collections;
using System.Collections.Generic;


namespace Loong.Game
{

    using UApp = Application;
    using NRB = NetworkReachability;

    /// <summary>
    /// 网络观察者
    /// </summary>
    public static class NetObserver
    {
        #region 字段
        /// <summary>
        /// 上次网络类型
        /// </summary>
        private static NRB last;

        /// <summary>
        /// 当前网络类型
        /// </summary>
        private static NRB cur;

        #endregion

        #region 属性
        /// <summary>
        /// 当前网络类型
        /// </summary>
        public static NRB Cur
        {
            get { return cur; }
        }

        #endregion

        #region 事件
        /// <summary>
        /// 网络类型发生改变时的事件 参数 <旧网络类型,当前网络类型>
        /// </summary>
        public static event Action<NRB, NRB> change = null;
        #endregion

        #region 构造方法
        static NetObserver()
        {
            last = Application.internetReachability;
            cur = Application.internetReachability;
        }
        #endregion

        #region 私有方法


        #endregion

        #region 保护方法

        #endregion

        #region 公开方法
        public static void Update()
        {
            cur = Application.internetReachability;
            if (last == cur) return;
            if (change != null)
            {
                change(last, cur);
				EventMgr.Trigger("NetChange", (int)last, (int)cur);
#if GAME_DEBUG
                iTrace.Warning("Loong", string.Format("nettype changed,before:{0},now:{1}", last, cur));
#endif
            }
            last = cur;
        }

        /// <summary>
        /// 移动数据
        /// </summary>
        /// <returns></returns>
        public static bool IsCarrier()
        {
            return (cur == NRB.ReachableViaCarrierDataNetwork);
        }

        /// <summary>
        /// 无网络
        /// </summary>
        /// <returns></returns>
        public static bool NoNet()
        {
            return (cur == NRB.NotReachable);
        }

        #endregion
    }
}